[Core] KISS 1 - Finite State Machine [Merge me first] - (Issue: #499) (#520)
pokt-network/pocket: state_machine/docs/README.md

├── module.go                     # Implementation of the StateMachine module
└── visualizer
    └── main.go                   # State machine diagram generator
```

## High Level Architecture

High-level implementation details:

- The [github.com/looplab/fsm](https://github.com/looplab/fsm) library is used to implement the FSM
- Pocket builds a wrapper around `looplab/fsm` to integrate with the other modules
- The `StateMachineModule` can be accessed via the `bus` from any other `IntegratableModule`
- State machine transitions emit `StateMachineTransitionEvent` events that subscribed pocket modules can listen to
- The `node` has a central [event handler](../../shared/node.go) for events that fan-out event handling to the relevant modules during state transitions

### FSM primer

The FSM has a declarative definition of an initial state and a set of transitions that have an `Event`, `Source` states and a single `Destination` state.

These are the main building blocks:

- **Event**: An event is a string that represents an action that can trigger a transition. For example, the event `start` can be used to trigger a transition from the `stopped` state to the `starting` state.

> - **State**: A state is a string that represents a state that the FSM can be in. For example, the state `stopped` can be used to represent a state where the node is not running.

Reviewer: But if the node is not running, does it have a state?

![Erwin_Schrödinger_(1933)](https://user-images.githubusercontent.com/1892194/219547569-79bee3b0-451d-44ba-bbcf-97395726b87c.jpg)

Author: LOL. 😍
Technically it would be in the initial state, by definition `stopped` right now

But for us to know that we would have to start it and then... we ... we ........... we ...................... [connection lost] 🤯

> - **Callback**: A callback is a function that is called when a transition occurs. For example, a callback can be used to log the transition or to perform some other action. (there are various types of callbacks that essentially drive the WHEN they are called and help building more complex behaviours like transition cancelling, etc. but we won't cover them here, please read the FSM library documentation for more details)

Reviewer: ```suggestion
- **Callback**: A callback is a function that is called when a transition occurs. For example, a callback can be used to log the transition or to perform some other action. Various types of callbacks essentially drive behaviour **WHEN** they are called and help build more complex behaviours like transition cancelling, etc. See the core FSM library documentation for more details.
```
